trino-s3-proxy/src/main/java/io/trino/s3/proxy/server/collections/MultiMap.java

```diff
@@ -0,0 +1,226 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.s3.proxy.server.collections;
+
+import com.google.common.collect.ImmutableSet;
+
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.BiConsumer;
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * A map between keys and values, each entry may have one or more values
+ * Null keys are not supported.
+ * The map may be configured to be case-insensitive at construction time, in which case
+ * all keys are lowercased.
+ *
+ * @param <V> type of the values
+ */
+public class MultiMap<V>
```

Comment:
The type parameter is always `String` right? Why not use `String` directly and not have a type parameter?